python/glow/wgr/linear_model/logistic_model.py

```diff
@@ -0,0 +1,245 @@
+                            sample_blocks: Dict[str, List[str]], modeldf: DataFrame,
+                            cvdf: DataFrame, covdf: pd.DataFrame, response: str) -> DataFrame:
+        """
+        Transforms a starting block matrix by applying a linear model.  The form of the output
+        can either be a direct linear transformation (response = "linear") or a linear transformation followed by a
+        sigmoid transformation (response = "sigmoid").
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            modeldf : Spark DataFrame produced by the LogisticRegression fit method, representing the reducer model
+            cvdf : Spark DataFrame produced by the LogisticRegression fit method, containing the results of the cross
+            validation routine.
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble.
+            response : String specifying what transformation to apply ("linear" or "sigmoid")
+
+        Returns:
+            Spark DataFrame containing the result of the transformation.
+        """
+
+        transform_key_pattern = ['sample_block', 'label']
+
+        if response == 'linear':
+            transform_udf = pandas_udf(
+                lambda key, pdf: apply_model(key, transform_key_pattern, pdf, labeldf,
+                                             sample_blocks, self.alphas, covdf),
+                reduced_matrix_struct, PandasUDFType.GROUPED_MAP)
+            join_type = 'inner'
+        elif response == 'sigmoid':
+            transform_udf = pandas_udf(
+                lambda key, pdf: apply_logistic_model(key, transform_key_pattern, pdf, labeldf,
+                                                      sample_blocks, self.alphas, covdf),
+                logistic_reduced_matrix_struct, PandasUDFType.GROUPED_MAP)
+            join_type = 'right'
+        else:
+            raise Exception(f'response must be either "linear" or "sigmoid", received "{response}"')
+
+        return blockdf.drop('header_block', 'sort_key') \
+            .join(modeldf.drop('header_block'), ['sample_block', 'header'], join_type) \
+            .withColumn('label', f.coalesce(f.col('label'), f.col('labels').getItem(0))) \
+            .groupBy(transform_key_pattern) \
+            .apply(transform_udf) \
+            .join(cvdf, ['label', 'alpha'], 'inner')
+
+    def transform(self, blockdf: DataFrame, labeldf: pd.DataFrame, sample_blocks: Dict[str,
+                                                                                       List[str]],
+                  modeldf: DataFrame, cvdf: DataFrame) -> pd.DataFrame:
```

Comment:
OOC, why we do not have covariates as an input to `transform` here?

Reply:
The transform method just returns the wgr term `X * b` from the model `logit(p) ~ C * a + X * b + g * e`, and does not include the covariates term, so covariates are not used here.